text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+
+The remaining behaviour of the function is equivalent to the `read_freeze` intrinsic.
+
+### Example Alternative Implementation
+
+If an implementation decides to define `MaybeUninit::freeze` (or equivalent) as the compiler intrinsic, it is possible to implement `core::ptr::read_freeze` as follows:
+```rust 
+pub unsafe fn read_freeze(ptr: *const T) -> T{
+    struct AlignedBytes([u8; core::mem::size_of::<T>()], [T;0]);
+    
+    let val = ptr.cast::<MaybeUninit<AlignedBytes>>().read();
+    
+    core::mem::transmute(val.freeze())
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawbacks that have been identified so far:
+* It is potentially [considered desireable](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Arguments.20for.20freeze/near/377333420) to maintain the property that sound (e.g. correct) code cannot meaningfully read uninitialized memory
+    * It is generally noted that safe/unsafe is not a security or privilege boundary, and it's fully possible to write unsound code (either deliberately or inadvertanly) that performs the read. If the use of uninitialized memory is within the threat model of a library that, for example, handles cryptographic secrets, that library should take additional steps to santize memory that contains those secrets.
+    * Undefined behaviour does not prevent malicious code from accessing any memory it physically can
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* The intrinsic could be defined as a by-value operation used by `MaybeUninit::freeze`, instead of the read operation used by `core::ptr::read_freeze`
+    * As noted above, it is believed that the two intrinsic choices are functionally identical, and thus the choice between them is completely arbitrary.
+* Either one of the two functions could be provided on their own
+    * Both functions are provided for maximum flexibility, and can be defined in terms of each other. The author does not believe there is significant drawback to providing both functions instead of just one
+* An in-place, mutable `freeze` could be offered, e.g. `MaybeUninit::freeze(&mut self)`
+    * While this function would seem to be a simple body that llvm could replace with a runtime no-op, in reality it is possible for [virtual memory](https://man7.org/linux/man-pages/man2/madvise.2.html#MADV_FREE) that has been freshly allocated and has not been written to exhibit properties of uninitialized memory (rather than simply being an abstract phenomenon the compiler tracks that disappears at runtime). Thus, such an operation would require a potentially expensive in-place copy. Until such time as an optimized version is available, we should avoid defining the in-place version, and require users to spell it explicitly as `*self = core::mem::replace(&mut self, uninit()).freeze()`.
```

Comment:
If we add this, the in-place version also can be implemented in a library crate, that uses inline asm to touch/write to each page at least once ensuring that it is not MADV_FREE.

Without the primitive freeze, there's no corresponding sequence of AM operations that this is equivalent to, but with it, the asm is equivalent to freezing each byte by-hand.

Reply:
Could you elaborate that a bit @thomcc? Like, what inline assembly are you thinking of? Because from your comment it feels like you're proposing both the use of ASM and the new operation.